File: backend/src/services/payments.rs
use sqlx::PgPool;
use uuid::Uuid;

use crate::errors::{AppError, AppResult};
use crate::models::payment::FeeBreakdown;

/// Create a payment record and Stripe PaymentIntent with manual capture.
/// Returns the PaymentIntent client_secret for frontend confirmation.
pub async fn create_escrow_payment(
    db: &PgPool,
    stripe_secret_key: &str,
    task_id: Uuid,
    requester_id: Uuid,
    doer_id: Uuid,
    task_price_cents: i64,
    requester_stripe_customer_id: &str,
) -> AppResult<(Uuid, String)> {
    let breakdown = FeeBreakdown::calculate(task_price_cents);

    let client = stripe::Client::new(stripe_secret_key);

    // Create PaymentIntent with manual capture (authorize only)
    let mut create_params =
        stripe::CreatePaymentIntent::new(breakdown.total_charged_cents, stripe::Currency::USD);
    create_params.capture_method = Some(stripe::PaymentIntentCaptureMethod::Manual);
    create_params.customer = Some(
        requester_stripe_customer_id
            .parse()
            .map_err(|_| AppError::PaymentError("Invalid Stripe customer ID".into()))?,
    );
    create_params.metadata = Some(
        [
            ("task_id".to_string(), task_id.to_string()),
            ("requester_id".to_string(), requester_id.to_string()),
            ("doer_id".to_string(), doer_id.to_string()),
        ]
        .into_iter()
        .collect(),
    );

    let payment_intent = stripe::PaymentIntent::create(&client, create_params)
        .await
        .map_err(|e| {
            AppError::PaymentError(format!("Stripe PaymentIntent creation failed: {e}"))
        })?;

    let payment_id = Uuid::now_v7();

    sqlx::query(
        r#"
        INSERT INTO payments
            (id, task_id, requester_id, doer_id, task_price_cents, gideon_fee_cents,
             stripe_fee_cents, total_charged_cents, doer_payout_cents,
             stripe_payment_intent_id, status, created_at)
        VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, 'pending', now())
        "#,
    )
    .bind(payment_id)
    .bind(task_id)
    .bind(requester_id)
    .bind(doer_id)
    .bind(breakdown.task_price_cents)
    .bind(breakdown.gideon_fee_cents)
    .bind(breakdown.stripe_fee_cents)
    .bind(breakdown.total_charged_cents)
    .bind(breakdown.doer_payout_cents)
    .bind(payment_intent.id.as_str())
    .execute(db)
    .await?;

    let client_secret = payment_intent
        .client_secret
        .ok_or_else(|| AppError::PaymentError("No client_secret on PaymentIntent".into()))?;

    Ok((payment_id, client_secret))
}

/// Capture an authorized PaymentIntent (transition from authorized → captured).
/// Called when the doer confirms start.
pub async fn capture_payment(db: &PgPool, stripe_secret_key: &str, task_id: Uuid) -> AppResult<()> {
    let payment: crate::models::payment::Payment =
        sqlx::query_as("SELECT * FROM payments WHERE task_id = $1")
            .bind(task_id)
            .fetch_optional(db)
            .await?
            .ok_or_else(|| AppError::NotFound("Payment not found".into()))?;

    let client = stripe::Client::new(stripe_secret_key);

    stripe::PaymentIntent::capture(
        &client,
        &payment.stripe_payment_intent_id,
        stripe::CapturePaymentIntent {
            ..Default::default()
        },
    )
    .await
    .map_err(|e| AppError::PaymentError(format!("Stripe capture failed: {e}")))?;

    sqlx::query("UPDATE payments SET status = 'escrowed', escrowed_at = now() WHERE task_id = $1")
        .bind(task_id)
        .execute(db)
        .await?;

    Ok(())
}

/// Release escrowed funds to the doer via Stripe Transfer.
/// Called when the requester approves completion.
pub async fn release_payment(db: &PgPool, stripe_secret_key: &str, task_id: Uuid) -> AppResult<()> {
    let payment: crate::models::payment::Payment =
        sqlx::query_as("SELECT * FROM payments WHERE task_id = $1")
            .bind(task_id)
            .fetch_optional(db)
            .await?
            .ok_or_else(|| AppError::NotFound("Payment not found".into()))?;

    // Get doer's Stripe Connect account
    let doer_connect_id: Option<String> =
        sqlx::query_scalar("SELECT stripe_connect_account_id FROM users WHERE id = $1")
            .bind(payment.doer_id)
            .fetch_optional(db)
            .await?
            .flatten();

    let doer_connect_id = doer_connect_id
        .ok_or_else(|| AppError::PaymentError("Doer has no Stripe Connect account".into()))?;

    let client = stripe::Client::new(stripe_secret_key);

    let mut transfer_params = stripe::CreateTransfer::new(stripe::Currency::USD, doer_connect_id);
    transfer_params.amount = Some(payment.doer_payout_cents);
    transfer_params.metadata = Some(
        [
            ("task_id".to_string(), task_id.to_string()),
            ("payment_id".to_string(), payment.id.to_string()),
        ]
        .into_iter()
        .collect(),
    );

    let transfer = stripe::Transfer::create(&client, transfer_params)
        .await
        .map_err(|e| AppError::PaymentError(format!("Stripe transfer failed: {e}")))?;

    sqlx::query(
        r#"
        UPDATE payments
        SET status = 'released', released_at = now(), stripe_transfer_id = $1
        WHERE task_id = $2
        "#,
    )
    .bind(transfer.id.as_str())
    .bind(task_id)
    .execute(db)
    .await?;

    Ok(())
}

/// Refund a payment (full refund of the PaymentIntent).
/// Gideon absorbs its own fee on refunds.
pub async fn refund_payment(db: &PgPool, stripe_secret_key: &str, task_id: Uuid) -> AppResult<()> {
    let payment: crate::models::payment::Payment =
        sqlx::query_as("SELECT * FROM payments WHERE task_id = $1")
            .bind(task_id)
            .fetch_optional(db)
            .await?
            .ok_or_else(|| AppError::NotFound("Payment not found".into()))?;

    let client = stripe::Client::new(stripe_secret_key);

    // If payment was only authorized (not captured), cancel the PaymentIntent
    if payment.status == "pending" {
        stripe::PaymentIntent::cancel(
            &client,
            &payment.stripe_payment_intent_id,
            stripe::CancelPaymentIntent {
                ..Default::default()
            },
        )
        .await
        .map_err(|e| AppError::PaymentError(format!("Stripe cancel failed: {e}")))?;
    } else {
        // Payment was captured — issue a refund
        let refund_params = stripe::CreateRefund {
            payment_intent: Some(
                payment
                    .stripe_payment_intent_id
                    .parse()
                    .map_err(|_| AppError::PaymentError("Invalid PaymentIntent ID".into()))?,
            ),
            ..Default::default()
        };

        stripe::Refund::create(&client, refund_params)
            .await
            .map_err(|e| AppError::PaymentError(format!("Stripe refund failed: {e}")))?;
    }

    sqlx::query("UPDATE payments SET status = 'refunded', refunded_at = now() WHERE task_id = $1")
        .bind(task_id)
        .execute(db)
        .await?;

    Ok(())
}

File: frontend/src/server/payments.ts
import "server-only";
import { execute, queryOne } from "./db";
import { newId } from "./auth";
import { calculateFees, feeBpsForVolume } from "./fees";
import { platformVolumeCents } from "./platform-volume";
import { stripe } from "./stripe";
import { paymentError, notFound } from "./errors";

interface PaymentRow {
  id: string;
  task_id: string;
  doer_id: string;
  task_price_cents: number;
  doer_payout_cents: number;
  stripe_payment_intent_id: string;
  stripe_transfer_id: string | null;
  status: string;
}

async function fetchPayment(taskId: string): Promise<PaymentRow> {
  const row = await queryOne<PaymentRow>(
    `SELECT id, task_id, doer_id, task_price_cents, doer_payout_cents,
            stripe_payment_intent_id, stripe_transfer_id, status
     FROM payments WHERE task_id = $1`,
    [taskId],
  );
  if (!row) throw notFound("Payment not found");
  return row;
}

export async function createEscrowPayment(
  taskId: string,
  requesterId: string,
  doerId: string,
  taskPriceCents: number,
  requesterStripeCustomerId: string,
): Promise<{ paymentId: string; clientSecret: string }> {
  const feeBps = feeBpsForVolume(await platformVolumeCents());
  const fees = calculateFees(taskPriceCents, feeBps);
  const s = stripe();

  let intent;
  try {
    intent = await s.paymentIntents.create({
      amount: fees.total_charged_cents,
      currency: "usd",
      capture_method: "manual",
      customer: requesterStripeCustomerId,
      metadata: {
        task_id: taskId,
        requester_id: requesterId,
        doer_id: doerId,
      },
    });
  } catch (e) {
    throw paymentError(`Stripe PaymentIntent creation failed: ${(e as Error).message}`);
  }

  if (!intent.client_secret) {
    throw paymentError("No client_secret on PaymentIntent");
  }

  const paymentId = newId();
  await execute(
    `INSERT INTO payments
       (id, task_id, requester_id, doer_id, task_price_cents, gideon_fee_cents,
        stripe_fee_cents, total_charged_cents, doer_payout_cents,
        stripe_payment_intent_id, status, created_at)
     VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, 'pending', now())`,
    [
      paymentId,
      taskId,
      requesterId,
      doerId,
      fees.task_price_cents,
      fees.gideon_fee_cents,
      fees.stripe_fee_cents,
      fees.total_charged_cents,
      fees.doer_payout_cents,
      intent.id,
    ],
  );

  return { paymentId, clientSecret: intent.client_secret };
}

export async function capturePayment(taskId: string): Promise<void> {
  const payment = await fetchPayment(taskId);
  try {
    await stripe().paymentIntents.capture(payment.stripe_payment_intent_id);
  } catch (e) {
    throw paymentError(`Stripe capture failed: ${(e as Error).message}`);
  }
  await execute(
    `UPDATE payments SET status = 'escrowed', escrowed_at = now() WHERE task_id = $1`,
    [taskId],
  );
}

export async function releasePayment(taskId: string): Promise<void> {
  const payment = await fetchPayment(taskId);
  const doerConnect = await queryOne<{ stripe_connect_account_id: string | null }>(
    `SELECT stripe_connect_account_id FROM users WHERE id = $1`,
    [payment.doer_id],
  );
  if (!doerConnect?.stripe_connect_account_id) {
    throw paymentError("Doer has no Stripe Connect account");
  }

  let transfer;
  try {
    transfer = await stripe().transfers.create({
      amount: payment.doer_payout_cents,
      currency: "usd",
      destination: doerConnect.stripe_connect_account_id,
      metadata: { task_id: taskId, payment_id: payment.id },
    });
  } catch (e) {
    throw paymentError(`Stripe transfer failed: ${(e as Error).message}`);
  }

  await execute(
    `UPDATE payments
     SET status = 'released', released_at = now(), stripe_transfer_id = $1
     WHERE task_id = $2`,
    [transfer.id, taskId],
  );
}

export async function refundPayment(taskId: string): Promise<void> {
  const payment = await fetchPayment(taskId);
  const s = stripe();

  if (payment.status === "pending") {
    try {
      await s.paymentIntents.cancel(payment.stripe_payment_intent_id);
    } catch (e) {
      throw paymentError(`Stripe cancel failed: ${(e as Error).message}`);
    }
  } else {
    try {
      await s.refunds.create({
        payment_intent: payment.stripe_payment_intent_id,
      });
    } catch (e) {
      throw paymentError(`Stripe refund failed: ${(e as Error).message}`);
    }
  }

  await execute(
    `UPDATE payments SET status = 'refunded', refunded_at = now() WHERE task_id = $1`,
    [taskId],
  );
}
